### ISSUE TITLE
about hyperparameters with transformer model in fairseq

### ISSUE THREAD

**Issue by weitaizhang** [pytorch/fairseq#187]
hi, @myleott ,
the new paper "scaling neural machine translation" speed up the transformer NMT with several ways. 
That's amazing and I am trying to reproduct the accuracy and efficiency with your methods.
could you please give some examples or command lines used in your experiments?Thx.

**Comment by myleott** [comment]
This is the basic command to replicate the results on 128 GPUs:
```
python train.py data-bin/wmt16_en_de_bpe32k/ \
  --fp16 \
  --arch transformer_vaswani_wmt_en_de_big --share-all-embeddings \
  --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
  --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 \
  --lr 0.001 --min-lr 1e-09 \
  --dropout 0.3 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
  --max-tokens 3584
```

Note that the `--fp16` flag requires you have CUDA 9.1 or greater and a Volta GPU. Also you should make sure you preprocessed your dataset after this fix: #180.

If you are training on fewer than 128 GPUs (let's say 8 GPUs), then you have a few options:
- reduce the learning rate: 0.0007 works well with 8 GPUs
- or: use `--update-freq 16` to simulate training on 8*16=128 GPUs

**Comment by weitaizhang** [comment]
@myleott wow, that's really nice of you .Thanks very much.

**Comment by myleott** [comment]
I've copied this here as well: https://github.com/pytorch/fairseq/tree/master/examples/translation#replicating-results-from-scaling-neural-machine-translation

**Comment by myleott** [comment]
It seems the Tensor Cores are not being activated. And because FP16 requires additional bookkeeping compares to FP32, this actually makes FP16 slower than FP32 (without Tensor Cores).

Can you make sure you have all the latest GEMM patches, and then reinstall PyTorch after applying the patches. The patches can be found here (depending on your platform): https://developer.nvidia.com/cuda-91-download-archive?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1704&target_type=runfilelocal

**Comment by myleott** [comment]
Hey, did you get this working? If not, please feel free to re-open (or open a new Issue) and we can help debug.

**Comment by weitaizhang** [comment]
@myleott yes, it's working now . Thank you very much ! Sorry I didn't give you feedback in time.

**Comment by AranKomat** [comment]
I'm sorry if this is irrelevant to the discussion. It seems that the input of fairseq is float32, not float16, while other mixed precision frameworks such as NVIDIA's Apex (Imagenet) uses float16 inputs. In both cases, the network has half-precision. Does this difference make no effect in practice?
